File: api/app/models/issue.rb
class Issue < ApplicationRecord
  include Uuidable

  STATUSES = {
    suggestion: 'suggestion',
    planned: 'planned',
    in_progress: 'in_progress',
    live: 'live'
  }.freeze

  belongs_to :user
  belongs_to :product
  belongs_to :issue_category
  has_many :comments, as: :parent, dependent: :destroy
  has_many :all_comments, class_name: 'Comment', dependent: :destroy
  has_many :issue_upvotes, dependent: :destroy

  validates :title, presence: true
  validates :detail, presence: true
  validates :status, presence: true, inclusion: { in: STATUSES.values }

  scope :latest_first, -> { order(created_at: :desc) }
  scope :sort_by_upvotes, -> (direction) { order(upvotes_count: direction) }
  scope :sort_by_comments, -> (direction) { order(comments_count: direction) }

  def upvoted_by?(user)
    issue_upvotes.exists?(user_id: user.id)
  end
end

File: web/src/feedbackr-api/v1/schemas/issue.ts
import z from 'zod'

import { IssueCategorySchema } from './issue-category'
import { UserSchema } from './user'
import { CommentSchema } from './comment'

export const IssueStatusSchema = z.enum(['suggestion', 'planned', 'in_progress', 'live'])

export type IssueStatus = z.infer<typeof IssueStatusSchema>

export const IssueSchema = z.object({
  uuid: z.string(),
  title: z.string(),
  detail: z.string(),
  status: IssueStatusSchema,
  createdAt: z.string(),
  updatedAt: z.string(),
  category: IssueCategorySchema,
  user: UserSchema,
  comments: z.array(CommentSchema).optional(),
  commentsCount: z.number(),
  upvotes: z.number(),
  alreadyUpvoted: z.boolean()
})

export type Issue = z.infer<typeof IssueSchema>
